lec03-basic-r.Rmd

```diff
@@ -1,4 +1,732 @@
+>  `hindfoot_half` containing values that are half the `hindfoot_length` values.
+>  In this `hindfoot_half` column, there are no `NA`s and all values are less
+>  than 30.
+>
+>  **Hint**: think about how the commands should be ordered to produce this data frame!
+
+<!---
+```{r, eval=FALSE}
+## Answer
+surveys_hindfoot_half <- surveys %>%
+    filter(!is.na(hindfoot_length)) %>%
+    mutate(hindfoot_half = hindfoot_length / 2) %>%
+    filter(hindfoot_half < 30) %>%
+    select(species_id, hindfoot_half)
+```
+--->
+
+
+### Split-apply-combine data analysis and the summarize() function
+
+Many data analysis tasks can be approached using the *split-apply-combine*
+paradigm: split the data into groups, apply some analysis to each group, and
+then combine the results. **`dplyr`** makes this very easy through the use of
+the `group_by()` function.
+
+
+#### The `summarize()` function
+
+`group_by()` is often used together with `summarize()`, which collapses each
+group into a single-row summary of that group.  `group_by()` takes as arguments
+the column names that contain the **categorical** variables for which you want
+to calculate the summary statistics. So to view the mean `weight` by sex:
+
+```{r}
+surveys %>%
+    group_by(sex) %>%
+    summarize(mean_weight = mean(weight, na.rm = TRUE))
+```
+
+You can also group by multiple columns:
+
+```{r}
+surveys %>%
+    group_by(sex, species_id) %>%
+    summarize(mean_weight = mean(weight, na.rm = TRUE))
+```
+
+When grouping both by `sex` and `species_id`, the first rows are for individuals
+that escaped before their sex could be determined and weighted. You may notice
+that the last column does not contain `NA` but `NaN` (which refers to "Not a
+Number", this is only visible in the interactive Notebook, not the knitted output). To avoid this, we can remove the missing values for weight before we
+attempt to calculate the summary statistics on weight. Because the missing
+values are removed, we can omit `na.rm = TRUE` when computing the mean:
+
+```{r}
+surveys %>%
+    filter(!is.na(weight)) %>%
+    group_by(sex, species_id) %>%
+    summarize(mean_weight = mean(weight))
+```
+
+Note that the number of not displayed rows decreased from 82 to 54. If you want
+to display more data, you can use the `print()` function at the end of your
+chain with the argument `n` specifying the number of rows to display:
+
+```{r}
+surveys %>%
+    filter(!is.na(weight)) %>%
+    group_by(sex, species_id) %>%
+    summarize(mean_weight = mean(weight)) %>%
+    print(n = 15)
+```
+
+Once the data are grouped, you can also summarize multiple variables at the same
+time (and not necessarily on the same variable). For instance, we could add a
+column indicating the minimum weight for each species for each sex:
+
+```{r}
+surveys %>%
+    filter(!is.na(weight)) %>%
+    group_by(sex, species_id) %>%
+    summarize(mean_weight = mean(weight),
+              min_weight = min(weight))
+```
+
+
+#### Tallying
```

Comment:
Maybe `## Using tally to summarize categorical data` or something a bit more informative.

Reply:
Reason I suggest *not* using `###` is because you use `###` for challenges, which makes it nicely separated on the TOC, so it's easy to quickly scan for challenges.

Reply:
Will either change this or make challenges `####`. I like having the challenges in the TOC so will switch back the previous lecture.

Reply:
Yes, I agree, Challenges in TOC are actually a good idea. Though I wouldn't recommend using `####` since it may not show up in the TOC without changing some options and a fourth level header just doesn't look good in the TOC.

Reply:
I would prefer the third level for the subsections as I have it right now, so I used fourth level for the challenges... It is not a big deal if they don't show up in the TOC, so up to you if you want to include 4th level headings or not.